skrub/_datetime_encoder.py

```diff
@@ -180,95 +170,145 @@ def fit(self, X: ArrayLike, y=None) -> "DatetimeEncoder":
             The data to transform, where each column is a datetime feature.
         y : None
             Unused, only here for compatibility.
 
         Returns
         -------
-        ndarray, shape (``n_samples``, ``n_features_out_``)
+        X_out : ndarray of shape ``(n_samples, n_features_out_)``
             Transformed input.
         """
-        check_is_fitted(
-            self,
-            attributes=["n_features_in_", "n_features_out_", "features_per_column_"],
-        )
-        X = check_input(X)
-        if X.shape[1] != self.n_features_in_:
-            raise ValueError(
-                f"The number of features in the input data ({X.shape[1]}) "
-                "does not match the number of features "
-                f"seen during fit ({self.n_features_in_}). "
-            )
-        # Create a new array with the extracted features,
-        # choosing only features that weren't constant during fit
-        X_ = np.empty((X.shape[0], self.n_features_out_), dtype=np.float64)
-        idx = 0
-        for i in range(X.shape[1]):
-            for j, feature in enumerate(self.features_per_column_[i]):
-                X_[:, idx + j] = self._extract_from_date(X[:, i], feature)
-            idx += len(self.features_per_column_[i])
-        return X_
-
-    def get_feature_names_out(self, input_features=None) -> list[str]:
-        """Return clean feature names.
+        check_is_fitted(self)
+        self._check_n_features(X, reset=False)
+        self._check_feature_names(X, reset=False)
+        X = check_array(X, ensure_2d=True, force_all_finite=False, dtype=None)
+
+        return self._parse_datetime_cols(X)
+
+    def _parse_datetime_cols(self, X):
+        """Extract datetime features from the selected columns.
+
+        Parameters
+        ----------
+        X : ndarray of shape ``(n_samples, n_features)``
+
+        Returns
+        -------
+        X_out : ndarray of shape ``(n_samples, n_features_out_)``
+        """
+        columns = getattr(self, "feature_names_in_", list(range(X.shape[1])))
+        # X_out must be of dtype float64 to handle np.nan
+        X_out = np.empty((X.shape[0], self.n_features_out_), dtype=np.float64)
+        offset_idx = 0
+        for col_idx, col in enumerate(columns):
+            if col in self.features_per_column_:
+                # X_col is a DatetimeIndex
+                X_col = pd.to_datetime(X[:, col_idx], errors=self.errors)
+
+                features = self.features_per_column_[col]
+                for feat_idx, feature in enumerate(features):
+                    if feature == "total_second":
+                        if X_col.tz is not None:
+                            X_col = X_col.tz_convert("utc")
+                        # Total seconds since epoch
+                        mask_notnull = X_col == X_col
+                        X_feature = np.where(
+                            mask_notnull,
+                            X_col.astype("int64") // 1e9,
```

Comment:
why use floor division?

Reply:
Useless here indeed, old reflex to get integers but `1e9` is a float and `X_out` dtype is also float